IdentityCore/src/cache/MSIDMacKeychainTokenCache.m

```diff
@@ -0,0 +1,666 @@
+                                                        serializer:serializer
+                                                           context:context
+                                                             error:error];
+
+    if (items.count > 1)
+    {
+        [self createError:@"The token cache store for this resource contains more than one user"
+                   domain:MSIDErrorDomain errorCode:MSIDErrorCacheMultipleUsers error:error context:context];
+        return nil;
+    }
+
+    return items.firstObject;
+}
+
+// Read one or more accounts from the keychain that match the key (see accountItem:matchesKey).
+// If not found, return an empty list without setting an error.
+//
+// Errors:
+// * MSIDKeychainErrorDomain/OSStatus: Apple status codes from SecItemCopyMatching()
+//
+- (NSArray<MSIDAccountCacheItem *> *)accountsWithKey:(MSIDCacheKey *)key
+                                          serializer:(id<MSIDAccountItemSerializer>)serializer
+                                             context:(id<MSIDRequestContext>)context
+                                               error:(NSError **)error
+{
+    MSID_TRACE;
+    MSID_LOG_VERBOSE(context, @"Get keychain items, key info (account: %@ service: %@ generic: %@ type: %@, keychainGroup: %@)", _PII_NULLIFY(key.account), key.service, _PII_NULLIFY(key.generic), key.type, [self keychainGroupLoggingName]);
+    MSID_LOG_VERBOSE_PII(context, @"Get keychain items, key info (account: %@ service: %@ generic: %@ type: %@, keychainGroup: %@)", key.account, key.service, key.generic, key.type, self.keychainGroup);
+
+    NSMutableDictionary *query = [self primaryAttributesForKey:key];
+    [query addEntriesFromDictionary:[self secondaryAttributesForKey:key]];
+    // Per Apple's docs, kSecReturnData can't be combined with kSecMatchLimitAll:
+    // https://developer.apple.com/documentation/security/1398306-secitemcopymatching?language=objc
+    // For this reason, we retrieve references to the items, then (below) use a second SecItemCopyMatching()
+    // to retrieve the data for each, deserializing each into an account object.
+    query[(id)kSecMatchLimit] = (id)kSecMatchLimitAll;
+    query[(id)kSecReturnRef] = @YES;
+
+    CFTypeRef cfItems = nil;
+    MSID_LOG_INFO(context, @"Trying to find keychain items...");
+    OSStatus status = SecItemCopyMatching((CFDictionaryRef)query, &cfItems);
+    MSID_LOG_INFO(context, @"Keychain find status: %d", (int)status);
+
+    if (status == errSecItemNotFound)
+    {
+        return @[];
+    }
+    else if (status != errSecSuccess)
+    {
+        [self createError:@"Failed to read account from keychain"
+                   domain:MSIDKeychainErrorDomain errorCode:status error:error context:context];
+        return nil;
+    }
+
+    NSArray *items = CFBridgingRelease(cfItems);
+
+    query = [self primaryAttributesForKey:key];
+    [query addEntriesFromDictionary:[self secondaryAttributesForKey:key]];
```

Comment:
nit: isn't this done already above? (reading the query and adding entries to it?)

Reply:
updated